### PR TITLE
Add hyperparameter search to Trainer

### PR DESCRIPTION
**Pull request by sgugger**
This PR introduces an access point to easily use optuna or Ray Tune hyperparameter search with `Trainer`.

Here is an example of use on MRPC (requires nlp installed from source on branch cloudpickle):
```
from nlp import load_dataset, load_metric
from transformers import AutoModelForSequenceClassification, AutoTokenizer, DataCollatorWithPadding, Trainer, TrainingArguments

tokenizer = AutoTokenizer.from_pretrained('bert-base-cased')
dataset = load_dataset('glue', 'mrpc')
metric = load_metric('glue', 'mrpc')

def encode(examples):
    outputs = tokenizer(examples['sentence1'], examples['sentence2'], truncation=True)
    return outputs

encoded_dataset = dataset.map(encode, batched=True)
# Won't be necessary when this PR is merged with master since the Trainer will do it automatically
encoded_dataset.set_format(columns=['attention_mask', 'input_ids', 'token_type_ids', 'label'])

def model_init():
    return AutoModelForSequenceClassification.from_pretrained('bert-base-cased', return_dict=True)

def compute_metrics(eval_pred):
    predictions, labels = eval_pred
    predictions = predictions.argmax(axis=-1)
    return metric.compute(predictions=predictions, references=labels)

# Evaluate during training and a bit more often than the default to be able to prune bad trials early.
# Disabling tqdm is a matter of preference.
training_args = TrainingArguments("test", evaluate_during_training=True, eval_steps=500, disable_tqdm=True)
trainer = Trainer(
    args=training_args,
    data_collator=DataCollatorWithPadding(tokenizer),
    train_dataset=encoded_dataset["train"], 
    eval_dataset=encoded_dataset["validation"], 
    model_init=model_init,
    compute_metrics=compute_metrics,
)

# Defaut objective is the sum of all metrics when metrics are provided, so we have to maximize it.
trainer.hyperparameter_search(direction="maximize")
```
This will use optuna or ray Tune depending on which platform is installed (default to optuna if both are there). You can force a backend by passing `backend="ray"` (or  `backend="optuna"`).

Here is another example on a simple regression problem (will convert to test in a follow-up PR):
```
from transformers import Trainer, TrainingArguments
import numpy as np

class RegressionDataset:
    def __init__(self, a=2, b=3, length=64, seed=42):
        np.random.seed(seed)
        self.length = length
        self.x = np.random.normal(size=(length,)).astype(np.float32)
        self.y = a * self.x + b + np.random.normal(scale=0.1, size=(length,)).astype(np.float32)
    
    def __len__(self):
        return self.length
    
    def __getitem__(self, i):
        return {'input_ids': self.x[i], 'label': self.y[i]}

class RegressionModel(torch.nn.Module):
    def __init__(self, a=0, b=0):
        super().__init__()
        self.a = torch.nn.Parameter(torch.tensor(a).float())
        self.b = torch.nn.Parameter(torch.tensor(b).float())
    
    def forward(self, input_ids=None, labels=None):
        y = input_ids * self.a + self.b
        if labels is None:
            return (y,)
        loss = torch.nn.functional.mse_loss(y, labels)
        return (loss, y)

train_set = RegressionDataset()
eval_set = RegressionDataset()

model = RegressionModel()
training_args = TrainingArguments("test", evaluate_during_training=True, eval_steps=8, disable_tqdm=True)

def compute_metrics(eval_pred):
    predictions, labels = eval_pred
    true = np.abs(predictions - labels) <= 0.25
    return {'accuracy': true.astype(np.float32).mean().item()}

trainer = Trainer(args=training_args, train_dataset=train_set, eval_dataset=eval_set, model_init=lambda: RegressionModel(), compute_metrics=compute_metrics)

best_trial = trainer.hyperparameter_search(direction="maximize")
```

To customize the hyperparameter searched, pass along a `hp_space` function like this (for optuna):
```
def my_hp_space(trial):
    return {
        "learning_rate": trial.suggest_float("learning_rate", 1e-2, 1, log=True),
        "num_train_epochs": trial.suggest_int("num_train_epochs", 1, 5),
        "seed": trial.suggest_int("seed", 1, 40),
        "per_device_train_batch_size": trial.suggest_categorical("per_device_train_batch_size", [4, 8, 16, 32, 64]),
    }
best_trial = trainer.hyperparameter_search(direction="maximize", hp_space=my_hp_space)
```

To customize the objective, pass along a `compute_objective` function.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -137,17 +140,21 @@ class Trainer:
...
+            The arguments to tweak training. Will default to a basic instance of :class:`~transformers.TrainingArguments`
+            with the ``output_dir`` set to a directory named `tmp_trainer` in the current directory if not provided.
         data_collator (:obj:`DataCollator`, `optional`, defaults to :func:`~transformers.default_data_collator`):
             The function to use to from a batch from a list of elements of :obj:`train_dataset` or
             :obj:`eval_dataset`.
         train_dataset (:obj:`torch.utils.data.dataset.Dataset`, `optional`):
             The dataset to use for training.
         eval_dataset (:obj:`torch.utils.data.dataset.Dataset`, `optional`):
             The dataset to use for evaluation.
+        model_init (:obj:`Callable[[], PreTrainedModel]`, `optional`):
```
Comment (sgugger) [inline src/transformers/trainer.py:155]: Adding this argument is necessary for hyperparameter search since we need to reinitialize the model before each new run.

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -113,6 +113,9 @@ class TrainingArguments:
             at the next training step under the keyword argument ``mems``.
         run_name (:obj:`str`, `optional`):
             A descriptor for the run. Notably used for wandb logging.
+        disable_tqdm (:obj:`bool`, `optional`):
```
Comment (sgugger) [inline src/transformers/training_args.py:116]: Adding this since all the progressbars while in hyperparameter search are overwhelming the output.

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -232,6 +235,13 @@ class TrainingArguments:
     run_name: Optional[str] = field(
         default=None, metadata={"help": "An optional descriptor for the run. Notably used for wandb logging."}
     )
+    disable_tqdm: Optional[str] = field(
```
Comment (julien-c) [inline src/transformers/training_args.py:238]:
```suggestion
    disable_tqdm: Optional[bool] = field(
```

File: src/transformers/trainer.py (hunk 1 above)
Comment (julien-c) [reply src/transformers/trainer.py:155]: Should we instead clone the `model` in each hp run? (~same memory/CPU hit than initializing a new one, I guess)

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -425,21 +442,60 @@ def setup_comet(self):
                 experiment._log_parameters(self.args, prefix="args/", framework="transformers")
                 experiment._log_parameters(self.model.config, prefix="config/", framework="transformers")
 
+    def setup_optuna(self, trial: "optuna.Trial"):
+        """
+        Setup the optional optuna integration. This is setting up good defaults for hyperparameter-search, subclass and
+        override to custoize the argumnents included in the search.
```
Comment (julien-c) [inline src/transformers/trainer.py:448]:
```suggestion
        override to customize the arguments included in the search.
```

File: src/transformers/trainer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/trainer.py:155]: You'd need to keep the copy for the next run, so that would take twice the space in memory.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -651,6 +714,51 @@ def train(self, model_path: Optional[str] = None):
...
+
+        Returns:
+            :class:`transformers.trainer_utils.BestRun`: All the informations about the best run.
+        """
+        assert is_optuna_available(), "optuna is required to do hyperparameter search, use `pip install optuna`."
+
+        def _objective(self, trial):
+            # To make sure optimizer and lr_scheduler are reset with the new choices of HPs
+            self.optimizer = None
+            self.lr_scheduler = None
```
Comment (julien-c) [inline src/transformers/trainer.py:748]: This means the `hyperparameter_search` is not compatible with passing a custom (optimizer, scheduler) to the Trainer, right?

File: src/transformers/trainer.py (hunk 1 above)
Comment (julien-c) [reply src/transformers/trainer.py:155]: Is the RAM (i.e. CPU RAM) footprint of the model usually a bottleneck in trainings? The trade off here is that you complexify the public API a little bit, which might or might not be warranted.

File: src/transformers/trainer.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/trainer.py:748]: Not for now. I don't think you can reset a scheduler in PyTorch so custom `(optimizer, scheduler)` means the user needs to subclass and override the `create_optimizer_and_scheduler` here. Will add to the documentation.

**Comment by julien-c** [comment]
Neat implem! I was initially wondering whether we should just have a `HyperTrainer` whose only job would be to spawn multiple instances of Trainers, given that the adherence between hp search and everything else in this class is rather small.

Thoughts? Others feel free to chime in

File: src/transformers/trainer.py (hunk 5 above)
Comment (julien-c) [reply src/transformers/trainer.py:748]: Yes, fair enough. I like subclassing as an extension point.

File: src/transformers/trainer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/trainer.py:155]: Some models take several GB and if the dataset is big, this is definitely going to bother the user. Another solution could be to save the initial model in a tempdir.

Note that I thought the API addition is valuable in itself IMO: I often want to compare some runs in a notebook and having two successive calls to train start from the initial model avoids reinstantiating it. But if it's just me, I can remove.

Comment (julien-c) [reply src/transformers/trainer.py:155]: Ok, that sounds reasonable.

Comment (thomwolf) [reply src/transformers/trainer.py:155]: Note that scanning the `seed` is important in Bert-like-model finetuning and means that we will actually want to have models with different random weights initialization.

**Review by sshleifer — APPROVED** [review]
This looks really cool. Excited for Seq2Seq integration!

Are hp results saved to disk by default?

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -431,15 +469,39 @@ def num_examples(self, dataloader: DataLoader) -> int:
...
+                The trial run or the hyperparameter dictionary for hyperparameter search.
         """
+        # Model re-init
+        if self.model_init is not None:
+            model = self.model_init()
+            self.model = model.to(self.args.device)
+
+        # HP search specific code
+        if self.hp_search_backend is not None and trial is not None:
+            config = self.hp_space(trial) if self.hp_search_backend == HPSearchBackend.OPTUNA else trial
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply src/transformers/trainer.py:490]: `params` might be more readable :)

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -651,6 +728,99 @@ def train(self, model_path: Optional[str] = None):
...
+                The number of trial runs to test.
+            direction(:obj:`str`, `optional`, defaults to :obj:`"minimize"`):
+                Whether to optimize greater or lower objects. Can be :obj:`"minimize"` or :obj:`"maximize"`, you should
+                pick :obj:`"minimize"` when optimizing the validation loss, :obj:`"maximize"` when optimizing one or
+                several metrics.
+            backend(:obj:`str` or :class:`~transformers.training_utils.HPSearchBackend`, `optional`):
+                The backend to use for hyperparameter search. Will default to optuna or Ray Tune, depending on which
+                one is installed. If both are installed, will default to optuna.
+            kwargs:
+                Additional keyword arguments passed along to :obj:`optuna.create_study` or :obj:`ray.tune.run`.
```
Comment (richardliaw) [inline src/transformers/trainer.py:765]: Maybe consider linking to the Ray/Optuna documentation for each of these calls too.

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -651,6 +728,99 @@ def train(self, model_path: Optional[str] = None):
...
+                    "At least one of optuna or ray should be installed. "
+                    "To install optuna run `pip install optuna`."
+                    "To install ray run `pip install ray[tune]`."
+                )
+        backend = HPSearchBackend(backend)
+        if backend == HPSearchBackend.OPTUNA and not is_optuna_available():
+            raise RuntimeError(" You picked the optuna backend, but it is not installed. Use `pip install optuna`.")
+        if backend == HPSearchBackend.RAY and not is_ray_available():
+            raise RuntimeError(
+                " You picked the Ray Tune backend, but it is not installed. Use `pip install ray[tune]`."
```
Comment (richardliaw) [inline src/transformers/trainer.py:783]:
```suggestion
                " You picked the Ray Tune backend, but it is not installed. Use `pip install 'ray[tune]'`."
```
^ to support zsh.

File: src/transformers/trainer.py [hunk 9]
```diff
@@ -651,6 +728,99 @@ def train(self, model_path: Optional[str] = None):
...
+            study.optimize(_objective, n_trials=n_trials, timeout=timeout, n_jobs=n_jobs)
+            best_trial = study.best_trial
+            best_run = BestRun(str(best_trial.number), best_trial.value, best_trial.params)
+        elif self.hp_search_backend == HPSearchBackend.RAY:
+            # The TensorBoard writer does not pickle so we have to remove it (if it exists) while doing the ray hp
+            # search.
+            _tb_writer = self.tb_writer
+            self.tb_writer = None
+            if "resources_per_trial" not in kwargs and self.args.n_gpu > 0:
+                kwargs["resources_per_trial"] = {"gpu": self.args.n_gpu}
```
Comment (richardliaw) [inline src/transformers/trainer.py:815]:
```suggestion
            if "resources_per_trial" not in kwargs and self.args.n_gpu > 0:
                kwargs["resources_per_trial"] = {"gpu": self.args.n_gpu}
            if "reporter" not in kwargs:
                 from ray.tune import CLIReporter
                 kwargs["reporter"] = CLIReporter(metric_columns=["objective"])
```
as mentioned in prior conversation :)

File: src/transformers/trainer.py [hunk 10]
```diff
@@ -651,6 +728,99 @@ def train(self, model_path: Optional[str] = None):
...
+        if backend == HPSearchBackend.RAY and not is_ray_available():
+            raise RuntimeError(
+                " You picked the Ray Tune backend, but it is not installed. Use `pip install ray[tune]`."
+            )
+        self.hp_search_backend = backend
+
+        self.hp_space = default_hp_space[backend] if hp_space is None else hp_space
+        self.compute_objective = default_compute_objective if compute_objective is None else compute_objective
+
+        def _objective(trial):
```
Comment (richardliaw) [inline src/transformers/trainer.py:790]: (out of scope) Just FYI, in the future, I think we will want to support something like population-based training, which requires an [additional argument `checkpoint_dir` to this signature](https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/pbt_transformers/pbt_transformers.py#L65), and a modification to [register saved state to Tune](https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/pbt_transformers/trainer.py#L48).

File: src/transformers/trainer.py [hunk 11]
```diff
@@ -651,6 +728,99 @@ def train(self, model_path: Optional[str] = None):
...
+            self.train(trial=trial)
+            # If there hasn't been any evaluation during the training loop.
+            if getattr(self, "objective", None) is None:
+                metrics = self.evaluate()
+                self.objective = self.compute_objective(metrics)
+            return self.objective
+
+        if self.hp_search_backend == HPSearchBackend.OPTUNA:
+            timeout = kwargs.pop("timeout", None)
+            n_jobs = kwargs.pop("n_jobs", 1)
```
Comment (richardliaw) [inline src/transformers/trainer.py:804]: I will make a PR on Ray side to also support this (or the equivalent)

File: src/transformers/trainer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LysandreJik) [reply src/transformers/trainer.py:490]: Would like `params` as well instead of `config`, which we use everywhere else for `PretrainedConfig` instances

File: src/transformers/trainer.py [hunk 12]
```diff
@@ -431,15 +469,39 @@ def num_examples(self, dataloader: DataLoader) -> int:
...
+        if self.model_init is not None:
+            model = self.model_init()
+            self.model = model.to(self.args.device)
+
+        # HP search specific code
+        if self.hp_search_backend is not None and trial is not None:
+            config = self.hp_space(trial) if self.hp_search_backend == HPSearchBackend.OPTUNA else trial
+            for key, value in config.items():
+                if not hasattr(self.args, key):
+                    logger.warn(
```
[the remark this answers is not in the document]
Comment (LysandreJik) [reply src/transformers/trainer.py:493]: I also think raising here would be better, there's usually a lot of logs in trainings and this might be unseen by the user.

File: src/transformers/training_args.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/training_args.py:116]: Will probably change the default to `None` once the logging PR is merged

**Review by LysandreJik — APPROVED** [review]
Great, looks good! Very cool feature!

File: src/transformers/trainer.py (hunk 11 above)
Comment (sgugger) [reply src/transformers/trainer.py:804]: That would be great!

**Review by LysandreJik — APPROVED** [review]
LGTM, great!

File: src/transformers/trainer.py (hunk 1 above)
Comment (aswincandra) [reply src/transformers/trainer.py:180]: Thankou so much! But I have a problem when defining the Trainer. It said, "__init__() got an unexpected keyword argument 'model_init'", what should I do? Is the Trainer doesn't detect 'model_init' parameter? Very sorry for the very newbie question :pray:t2: and Thankyou before.
